lib/Sema/CSRanking.cpp

```diff
@@ -1381,3 +1353,61 @@ SolutionDiff::SolutionDiff(ArrayRef<Solution> solutions) {
     }
   }
 }
+
+InputMatcher::InputMatcher(const ValueDecl *funcDecl,
+    const ArrayRef<AnyFunctionType::Param> params) :
+    NumSkippedParameters(0), Params(params) {
+  DefaultValueMap = computeDefaultMap(params, funcDecl,
+      funcDecl->getDeclContext()->isTypeContext());
+}
+
+InputMatcher::Result InputMatcher::solve(int numInputs,
+    bool ignoreLastPair,
```

Comment:
It looks like `ignoreLastPair` parameter is something specific to ranking it probably shouldn't be specially cased in the matcher itself, because it could be modeled outside by callee itself.

Reply:
Agree. I removed such `ignoreLastPair`. In this case, the parameters should be properly prepared from outside by caller. If we intend to ignore trailing closure, we can slice `params` by dropping the last element like [this](https://github.com/dingobye/swift/commit/da2b9d3a39ad86ac02dc47ad075c10839792d13b#diff-a72024f35020a11c7f1e09bcd35b22bdR664), before passing it to `InputMatcher`'s constructor.

As a side effect, however, `computeDefaultMap` does not produce expected results if we pass the sliced `params` to it. It means that we cannot directly compute `defaultMap` inside `InputMatcher`. As a result, `defaultMap` should be prepared outside by caller as well. So I added the corresponding updates.